parlai/tasks/squad2/agents.py

```diff
@@ -0,0 +1,391 @@
+                for qa in paragraph['qas']:
+                    question = qa['question']
+                    ans_iter = [{"text": ''}]
+                    if not qa['is_impossible']:
+                        ans_iter = qa['answers']
+                    answers = (a['text'] for a in ans_iter)
+                    yield (question, answers), True
+
+
+class TitleTeacher(DefaultTeacher):
+    """This version of SquAD inherits from the Default Teacher. The only
+    difference is that the 'text' field of an observation will contain
+    the title of the article separated by a newline from the paragraph and the
+    query.
+    Note: The title will contain underscores, as it is the part of the link for
+    the Wikipedia page; i.e., the article is at the site:
+    https://en.wikipedia.org/wiki/{TITLE}
+    Depending on your task, you may wish to remove underscores.
+    """
+
+    def __init__(self, opt, shared=None):
+        self.id = 'squad_title'
+        build(opt)
+        super().__init__(opt, shared)
+
+    def setup_data(self, path):
+        print('loading: ' + path)
+        with open(path) as data_file:
+            self.squad = json.load(data_file)['data']
+        for article in self.squad:
+            title = article['title']
+            # each paragraph is a context for the attached questions
+            for paragraph in article['paragraphs']:
+                # each question is an example
+                for qa in paragraph['qas']:
+                    question = qa['question']
+                    ans_iter = [{"text": ""}]
+                    if not qa['is_impossible']:
+                        ans_iter = qa['answers']
+                    answers = (a['text'] for a in ans_iter)
+                    context = paragraph['context']
+                    yield (
+                        '\n'.join([title, context, question]),
+                        answers
+                    ), True
+
+
+class SentenceIndexTeacher(IndexTeacher):
+    """Index teacher where the labels are the sentences the contain the true
+    answer.
+    """
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+
+        try:
+            import nltk
+        except ImportError:
+            raise ImportError('Please install nltk (e.g. pip install nltk).')
+        # nltk-specific setup
+        st_path = 'tokenizers/punkt/{0}.pickle'.format('english')
+        try:
+            self.sent_tok = nltk.data.load(st_path)
+        except LookupError:
+            nltk.download('punkt')
+            self.sent_tok = nltk.data.load(st_path)
+
+    def get(self, episode_idx, entry_idx=None):
+        article_idx, paragraph_idx, qa_idx = self.examples[episode_idx]
+        article = self.squad[article_idx]
+        paragraph = article['paragraphs'][paragraph_idx]
+        qa = paragraph['qas'][qa_idx]
+        context = paragraph['context']
+        question = qa['question']
+
+        answers = ['']
```

Comment:
here we'll want to start with an empty list, as `'' in 'any string'` returns true, which would mark every sentence as containing the label in this case